### ISSUE TITLE
app-static-to-dynamic-error with generateMetadata using getI18n on page not found

### ISSUE THREAD

**Issue by AvaN0x** [QuiiBz/next-international#298]
**Describe the bug**

Hello,
I get a [app-static-to-dynamic-error](https://nextjs.org/docs/messages/app-static-to-dynamic-error) error when I go in a route that is not matched by the middleware.

**To Reproduce**
1. Clone this repro repo : https://github.com/AvaN0x/repro-next-international-generate-metadata
1. 
```sh
npm i
npm run build
npm run start
```
3. Click on the link in the page
1. See the following error:
```log
 ⨯ Page changed from static to dynamic at runtime /a.a, reason: headers
see more here https://nextjs.org/docs/messages/app-static-to-dynamic-error
```
![image](https://github.com/QuiiBz/next-international/assets/27494805/9c6de699-a38b-4424-8ce2-4c92107dc515)
![image](https://github.com/QuiiBz/next-international/assets/27494805/db64d2bb-08f3-480e-826f-c90af4a7bd5c)

I also see that, when the path do not match the middleware, it'll still use the first part as the locale, this can be seein with the log `locale not found a.a` in the console.

**Expected behavior**
I'd expect the not-found page to show, instead I get a 500 error.
The error disappear when you remove the use of `getI18n` in the `generateMetadata` function in `[locale]/layout.ts`.
The error does not occur in `npm run dev`.

**About (please complete the following information):**
 - next-international version : 1.1.4
 - Next.js version: 14.0.3

**Infos about the repro**
- it was created using `npx create-next-app@latest`
- it contains a full next-international config
- it has a button to change locale, even if it's not needed for the repro, it might still be useful
- it also has a check in `[locale]/layout.ts` RootLayout to know if the locale params is indeed a locale or something else

I can provide more infos if needed

**Comment by AvaN0x** [comment]
I found a fix to the issue (*the repro has been updated accordingly*) with a call to `setStaticParamsLocale` in the `generateMetadata` function.
This is related to this part of the code that'll check the headers if no staticParams were found:
https://github.com/QuiiBz/next-international/blob/e4acd2fdf3c6ec181e218df82477f05df90c7311/packages/next-international/src/app/server/get-locale-cache.tsx#L19-L31

---

But I still get `locale not found a.a` logs in the layout by clicking the link in the repro.
It would be nice if the `params.locale` value was always a valid locale.

Because of that, I have to add a check before passing the value to `setStaticParamsLocale` or redirect myself to the not found page.

https://github.com/AvaN0x/repro-next-international-generate-metadata/blob/80920471993de4b470e076d99254025e3abdff14/src/app/%5Blocale%5D/layout.tsx#L16-L18

https://github.com/AvaN0x/repro-next-international-generate-metadata/blob/80920471993de4b470e076d99254025e3abdff14/src/app/%5Blocale%5D/layout.tsx#L42-L46

`locales` and `fallbackLocale` are coming from my `locales/shared.ts` file.

**Comment by QuiiBz** [comment]
> It would be nice if the params.locale value was always a valid locale.

Agree, that's why we use `generateStaticParams` to make sure the dynamic params are valid. This again seems like a bug with Next.js, I'll have to dig more.
